lib/spack/spack/solver/asp.py

```diff
@@ -2043,34 +2043,45 @@ def package_splice_rules(self, pkg):
 
             self.gen.newline()
 
-    def virtual_preferences(self, pkg_name, func):
-        """Call func(vspec, provider, i) for each of pkg's provider prefs."""
-        config = spack.config.get("packages")
-        pkg_prefs = config.get(pkg_name, {}).get("providers", {})
-        for vspec, providers in pkg_prefs.items():
-            if vspec not in self.possible_virtuals:
-                continue
+    def virtual_requirements_and_weights(self):
+        virtual_preferences = spack.config.CONFIG.get("packages:all:providers", {})
 
-            for i, provider in enumerate(providers):
-                provider_name = spack.spec.Spec(provider).name
-                func(vspec, provider_name, i)
+        self.gen.h1("Virtual requirements and weights")
+        for virtual_str in sorted(self.possible_virtuals):
             self.gen.newline()
+            self.gen.h2(f"Virtual: {virtual_str}")
+            self.gen.fact(fn.virtual(virtual_str))
 
-    def provider_defaults(self):
-        self.gen.h2("Default virtual providers")
-        self.virtual_preferences(
-            "all", lambda v, p, i: self.gen.fact(fn.default_provider_preference(v, p, i))
-        )
-
-    def provider_requirements(self):
-        self.gen.h2("Requirements on virtual providers")
-        for virtual_str in sorted(self.possible_virtuals):
             rules = self.requirement_parser.rules_from_virtual(virtual_str)
+            if not rules and virtual_str not in virtual_preferences:
+                continue
+
+            required, preferred = [], []
+            for rule in rules:
+                # We don't deal with conditional requirements
+                if rule.condition != spack.spec.Spec():
+                    continue
+
+                if len(rule.requirements) == 1:
+                    required.append(rule.requirements[0].name)
+                    continue
+
+                preferred.extend([x.name for x in rule.requirements if x.name])
```

Comment:
```suggestion
                else:
                    preferred.extend(x.name for x in rule.requirements if x.name)
```

More readable to me this way, given that "required" is not really an early exit condition, but one of the two outcomes of the loop.